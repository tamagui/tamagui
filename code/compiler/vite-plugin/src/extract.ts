// fork from https://github.com/seek-oss/vanilla-extract

import type { TamaguiOptions } from '@tamagui/static'
import path from 'node:path'
import type { Plugin, ResolvedConfig, ViteDevServer } from 'vite'
import { normalizePath, type Environment } from 'vite'
import {
  Static,
  disableStatic,
  extractor,
  loadTamaguiBuildConfig,
  tamaguiOptions,
} from './loadTamagui'

import { createHash } from 'node:crypto'

export function tamaguiExtractPlugin(optionsIn?: Partial<TamaguiOptions>): Plugin {
  if (optionsIn?.disable) {
    return {
      name: 'tamagui-extract',
    }
  }

  const getHash = (input: string) => createHash('sha1').update(input).digest('base64')

  const clearCompilerCache = () => {
    memoryCache = {}
    cacheSize = 0
  }

  let memoryCache = {}
  let cacheSize = 0

  const cssMap = new Map<string, string>()

  let config: ResolvedConfig
  let server: ViteDevServer
  const virtualExt = `.tamagui.css`

  const getAbsoluteVirtualFileId = (filePath: string) => {
    if (filePath.startsWith(config.root)) {
      return filePath
    }
    return normalizePath(path.join(config.root, filePath))
  }

  function isVite6AndNotClient(environment?: Environment) {
    return environment?.name && environment.name !== 'client'
  }

  function isVite6Native(environment?: Environment) {
    return (
      environment?.name && (environment.name === 'ios' || environment.name === 'android')
    )
  }

  function invalidateModule(absoluteId: string) {
    if (!server) return

    const { moduleGraph } = server
    const modules = moduleGraph.getModulesByFile(absoluteId)

    if (modules) {
      for (const module of modules) {
        moduleGraph.invalidateModule(module)

        // Vite uses this timestamp to add `?t=` query string automatically for HMR.
        module.lastHMRTimestamp = module.lastInvalidationTimestamp || Date.now()
      }
    }
  }

  return {
    name: 'tamagui-extract',
    enforce: 'pre',

    configureServer(_server) {
      server = _server
    },

    async buildStart() {
      await loadTamaguiBuildConfig(optionsIn)
    },

    buildEnd() {
      extractor?.cleanupBeforeExit()
    },

    config(userConf) {
      userConf.optimizeDeps ||= {}
      userConf.optimizeDeps.include ||= []
      userConf.optimizeDeps.include.push('@tamagui/core/inject-styles')
    },

    async configResolved(resolvedConfig) {
      config = resolvedConfig
    },

    async resolveId(source) {
      if (isVite6Native(this.environment)) {
        return
      }

      if (
        tamaguiOptions?.disableServerOptimization &&
        isVite6AndNotClient(this.environment)
      ) {
        // only optimize on client - server should produce identical styles anyway!
        return
      }

      const [validId, query] = source.split('?')

      if (!validId.endsWith(virtualExt)) {
        return
      }

      // Absolute paths seem to occur often in monorepos, where files are
      // imported from outside the config root.
      const absoluteId = source.startsWith(config.root)
        ? source
        : getAbsoluteVirtualFileId(validId)

      // There should always be an entry in the `cssMap` here.
      // The only valid scenario for a missing one is if someone had written
      // a file in their app using the .tamagui.js/.tamagui.css extension
      if (cssMap.has(absoluteId)) {
        // Keep the original query string for HMR.
        return absoluteId + (query ? `?${query}` : '')
      }
    },

    /**
     * TODO
     *
     *   mainFields module:jsx breaks, so lets just have a mapping here
     *   where we load() and map it to the jsx path before transform
     *
     */

    async load(id) {
      if (disableStatic) {
        // only optimize on client - server should produce identical styles anyway!
        return
      }

      if (isVite6Native(this.environment)) {
        return
      }

      if (
        tamaguiOptions?.disableServerOptimization &&
        isVite6AndNotClient(this.environment)
      ) {
        return
      }

      const [validId] = id.split('?')
      return cssMap.get(validId)
    },

    transform: {
      order: 'pre',
      async handler(code, id) {
        if (disableStatic) {
          // only optimize on client - server should produce identical styles anyway!
          return
        }

        if (isVite6Native(this.environment)) {
          return
        }

        if (
          tamaguiOptions?.disableServerOptimization &&
          isVite6AndNotClient(this.environment)
        ) {
          return
        }

        const [validId] = id.split('?')
        if (!validId.endsWith('.tsx')) {
          return
        }

        const firstCommentIndex = code.indexOf('// ')
        const { shouldDisable, shouldPrintDebug } = Static!.getPragmaOptions({
          source: firstCommentIndex >= 0 ? code.slice(firstCommentIndex) : '',
          path: validId,
        })

        if (shouldPrintDebug) {
          console.trace(
            `Current file: ${id} in environment: ${this.environment?.name}, shouldDisable: ${shouldDisable}`
          )
          console.info(`\n\nOriginal source:\n${code}\n\n`)
        }

        if (shouldDisable) {
          return
        }

        const cacheEnv =
          this.environment.name === 'client' || this.environment.name === 'ssr'
            ? // same cache key for ssr and web since they are the same
              'web'
            : this.environment.name
        const cacheKey = getHash(`${cacheEnv}${code}${id}`)
        const cached = memoryCache[cacheKey]
        if (cached) {
          return cached
        }

        const extracted = await Static!.extractToClassNames({
          extractor: extractor!,
          source: code,
          sourcePath: validId,
          options: tamaguiOptions!,
          shouldPrintDebug,
        })

        if (!extracted) {
          return
        }

        const rootRelativeId = `${validId}${virtualExt}`
        const absoluteId = getAbsoluteVirtualFileId(rootRelativeId)

        let source = extracted.js

        if (extracted.styles) {
          this.addWatchFile(rootRelativeId)

          if (server && cssMap.has(absoluteId)) {
            invalidateModule(rootRelativeId)
          }

          source = `${source}\nimport "${rootRelativeId}";`
          cssMap.set(absoluteId, extracted.styles)
        }

        const codeOut = source.toString()
        const out = {
          code: codeOut,
          map: extracted.map,
        }

        cacheSize += codeOut.length
        // ~50Mb cache for recently compiler files
        if (cacheSize > 26214400) {
          clearCompilerCache()
        }
        memoryCache[cacheKey] = out

        return out
      },
    },
  }
}
