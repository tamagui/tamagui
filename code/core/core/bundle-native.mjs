#!/usr/bin/env node

import { bundleNative } from '@tamagui/native-bundle'

// Bundle for production (native.cjs)
await bundleNative({
  entry: 'src/index.tsx',
  outDir: 'dist',
  fileName: 'native.cjs',
})

// Bundle for tests (test.native.cjs)
await bundleNative({
  entry: 'src/index.tsx',
  outDir: 'dist',
  fileName: 'test.native.cjs',
  define: {
    'process.env.NODE_ENV': JSON.stringify('test'),
  },
})
