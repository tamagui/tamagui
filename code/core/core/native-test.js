Object.assign(module.exports, require('./dist/test'))
