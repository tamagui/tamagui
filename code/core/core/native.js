Object.assign(module.exports, require('./dist/native.cjs'))
