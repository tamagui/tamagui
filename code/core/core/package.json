{
  "name": "@tamagui/core",
  "version": "1.125.8",
  "source": "src/index.tsx",
  "main": "dist/cjs",
  "module": "dist/esm",
  "tamagui": {
    "bundle.native": "./src/index.tsx",
    "bundle.native.test": "./src/index.tsx"
  },
  "types": "./types/index.d.ts",
  "license": "MIT",
  "author": {
    "name": "Nate Wienert"
  },
  "sideEffects": [
    "*.css"
  ],
  "scripts": {
    "build": "tamagui-build",
    "watch": "tamagui-build --watch",
    "lint": "biome check src",
    "lint:fix": "biome check --write src",
    "clean": "tamagui-build clean",
    "clean:build": "tamagui-build clean:build"
  },
  "files": [
    "types",
    "dist",
    "src",
    "reset.css",
    "native.js",
    "native.d.ts",
    "native-test.js",
    "native-test.d.ts"
  ],
  "dependencies": {
    "@tamagui/react-native-media-driver": "workspace:*",
    "@tamagui/react-native-use-pressable": "workspace:*",
    "@tamagui/react-native-use-responder-events": "workspace:*",
    "@tamagui/use-event": "workspace:*",
    "@tamagui/web": "workspace:*"
  },
  "devDependencies": {
    "@tamagui/build": "workspace:*",
    "@testing-library/react": "^16.1.0",
    "csstype": "^3.0.10",
    "typescript": "^5.7.2",
    "vitest": "^3.0.5"
  },
  "exports": {
    "./package.json": "./package.json",
    "./reset.css": "./reset.css",
    ".": {
      "react-native": {
        "import": "./dist/esm/index.native.js",
        "require": "./dist/cjs/index.native.js"
      },
      "types": "./types/index.d.ts",
      "import": "./dist/esm/index.mjs",
      "require": "./dist/cjs/index.cjs",
      "default": "./dist/cjs/index.native.js"
    },
    "./native": {
      "types": "./types/index.d.ts",
      "import": "./dist/native.mjs",
      "require": "./dist/native.js",
      "react-native": "./dist/native.js"
    },
    "./native-test": {
      "types": "./types/index.d.ts",
      "import": "./dist/test.native.js",
      "require": "./dist/test.native.js",
      "react-native": "./dist/test.native.js"
    },
    "./inject-styles": {
      "types": "./types/inject-styles.d.ts",
      "import": "./dist/esm/inject-styles.mjs",
      "require": "./dist/cjs/inject-styles.cjs",
      "react-native": "./dist/cjs/inject-styles.native.js"
    },
    "./reactNativeTypes": {
      "types": "./types/reactNativeTypes.d.ts",
      "import": "./dist/esm/reactNativeTypes.mjs",
      "require": "./dist/cjs/reactNativeTypes.cjs",
      "react-native": "./dist/cjs/reactNativeTypes.native.js"
    }
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/tamagui/tamagui.git",
    "directory": "packages/core"
  },
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "a49cc7ea6b93ba384e77a4880ae48ac4a5635c14"
}
