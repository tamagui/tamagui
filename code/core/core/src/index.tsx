import { createMedia } from '@tamagui/react-native-media-driver'
import { useResponderEvents } from '@tamagui/react-native-use-responder-events'
import type {
  StackNonStyleProps,
  StackStyleBase,
  TamaDefer,
  TamaguiComponent,
  TamaguiElement,
  TamaguiTextElement,
  TextNonStyleProps,
  TextProps,
  TextStylePropsBase,
} from '@tamagui/web'
import {
  Stack as WebStack,
  Text as WebText,
  View as WebView,
  composeEventHandlers,
  createTamagui as createTamaguiWeb,
  setupHooks,
} from '@tamagui/web'
import React from 'react'

import { createOptimizedView } from './createOptimizedView'
import { getBaseViews } from './getBaseViews'
import { getRect } from './helpers/getRect'
import { measureLayout, useElementLayout } from './hooks/useElementLayout'
import type { RNTextProps, RNViewProps } from './reactNativeTypes'
import { usePressability } from './vendor/Pressability'
import { addNativeValidStyles } from './addNativeValidStyles'

// helpful for usage outside of tamagui
export {
  getBoundingClientRectAsync,
  measureLayout,
  useElementLayout,
  type LayoutEvent,
} from './hooks/useElementLayout'

// add newer style props based on react native version
addNativeValidStyles()

// adds extra types to View/Stack/Text:

type RNExclusiveViewProps = Omit<RNViewProps, keyof StackNonStyleProps>
export interface RNTamaguiViewNonStyleProps
  extends StackNonStyleProps,
    RNExclusiveViewProps {}

type RNTamaguiView = TamaguiComponent<
  TamaDefer,
  TamaguiElement,
  RNTamaguiViewNonStyleProps,
  StackStyleBase,
  {}
>

type RNExclusiveTextProps = Omit<RNTextProps, keyof TextProps>
export interface RNTamaguiTextNonStyleProps
  extends TextNonStyleProps,
    RNExclusiveTextProps {}

type RNTamaguiText = TamaguiComponent<
  TamaDefer,
  TamaguiTextElement,
  RNTamaguiTextNonStyleProps,
  TextStylePropsBase,
  {}
>

// re-exports all of @tamagui/web just adds hooks
export * from '@tamagui/web'
// fixes issues with TS saying internal type usage is breaking
// see https://discord.com/channels/909986013848412191/1146150253490348112/1146150253490348112
export * from './reactNativeTypes'

// automate using the react native media driver
export const createTamagui: typeof createTamaguiWeb = (conf) => {
  if (process.env.TAMAGUI_TARGET === 'native') {
    if (conf.media) {
      conf.media = createMedia(conf.media)
    }
  }
  return createTamaguiWeb(conf)
}

const baseViews = getBaseViews()

// setup internal hooks:

setupHooks({
  getBaseViews,

  setElementProps: (node) => {
    // web only
    if (node && !node['measure']) {
      // @ts-ignore
      node.measure ||= (callback) => measureLayout(node, null, callback)
      // @ts-ignore
      node.measureLayout ||= (relativeToNode, success) =>
        measureLayout(node as HTMLElement, relativeToNode, success)
      // @ts-ignore
      node.measureInWindow ||= (callback) => {
        setTimeout(() => {
          const { height, left, top, width } = getRect(node as HTMLElement)!
          callback(left, top, width, height)
        }, 0)
      }
    }
  },

  usePropsTransform(elementType, propsIn, stateRef, willHydrate) {
    if (process.env.TAMAGUI_TARGET === 'web') {
      const isDOM = typeof elementType === 'string'

      // replicate react-native-web functionality
      const {
        // remove event props handles by useResponderEvents
        onMoveShouldSetResponder,
        onMoveShouldSetResponderCapture,
        onResponderEnd,
        onResponderGrant,
        onResponderMove,
        onResponderReject,
        onResponderRelease,
        onResponderStart,
        onResponderTerminate,
        onResponderTerminationRequest,
        onScrollShouldSetResponder,
        onScrollShouldSetResponderCapture,
        onSelectionChangeShouldSetResponder,
        onSelectionChangeShouldSetResponderCapture,
        onStartShouldSetResponder,
        onStartShouldSetResponderCapture,

        // android
        collapsable,
        focusable,

        // deprecated,
        accessible,
        accessibilityDisabled,

        onLayout,
        hrefAttrs,

        ...plainDOMProps
      } = propsIn

      if (willHydrate || isDOM) {
        useElementLayout(stateRef, !isDOM ? undefined : (onLayout as any))
        useResponderEvents(stateRef, !isDOM ? undefined : propsIn)
      }

      if (isDOM) {
        // TODO move into getSplitStyles
        if (plainDOMProps.href && hrefAttrs) {
          const { download, rel, target } = hrefAttrs
          if (download != null) {
            plainDOMProps.download = download
          }
          if (rel) {
            plainDOMProps.rel = rel
          }
          if (typeof target === 'string') {
            plainDOMProps.target = target.charAt(0) !== '_' ? `_${target}` : target
          }
        }
        return plainDOMProps
      }
    }
  },

  useEvents(viewProps, events, { pseudos }, setStateShallow, staticConfig) {
    if (process.env.TAMAGUI_TARGET === 'native') {
      if (events) {
        if (events.onFocus) {
          viewProps['onFocus'] = events.onFocus
        }
        if (events.onBlur) {
          viewProps['onBlur'] = events.onBlur
        }
      }

      if (staticConfig.isInput) {
        if (events) {
          const { onPressIn, onPressOut, onPress } = events
          const inputEvents = {
            onPressIn,
            onPressOut: onPressOut || onPress,
          }
          if (onPressOut && onPress) {
            // only supports onPressIn and onPressOut so combine them
            inputEvents.onPressOut = composeEventHandlers(onPress, onPressOut)
          }
          Object.assign(viewProps, inputEvents)
        }
      } else {
        // use Pressability to get smooth unPress when you press + hold + move out
        // only ever create once, use .configure() to update later
        if (events && viewProps.hitSlop) {
          events.hitSlop = viewProps.hitSlop
        }

        // note we do events checks more than we should because we need this hook to always run
        const pressability = usePressability(events)

        if (events) {
          if (process.env.NODE_ENV === 'development') {
            if (viewProps['debug']) {
              console.info(
                `Checking for press ${!!events.onPress} then applying pressability props: ${Object.keys(
                  pressability || {}
                )}`
              )
            }
          }

          if (events.onPress) {
            for (const key in pressability) {
              const og = viewProps[key]
              const val = pressability[key]
              viewProps[key] =
                og && !dontComposePressabilityKeys[key]
                  ? composeEventHandlers(og, val)
                  : val
            }
          }
        }
      }
    }
  },

  // attempt at properly fixing RN input, but <Pressable><TextInput /> just doesnt work on RN
  ...(process.env.TAMAGUI_TARGET === 'native' && {
    useChildren(elementType, children, viewProps) {
      if (process.env.NODE_ENV === 'test') {
        // test mode - just use regular views since optimizations cause weirdness
        return
      }

      if (elementType === baseViews.View) {
        // optimize view
        return createOptimizedView(children, viewProps, baseViews)
      }

      if (process.env.TAMAGUI_OPTIMIZE_NATIVE_VIEWS) {
        if (elementType === baseViews.Text) {
          // further optimize by not even caling elementType.render
          viewProps.children = children
          return React.createElement('RCTText', viewProps)
        }
      }
    },
  }),
})

const dontComposePressabilityKeys = {
  onClick: true,
}

// overwrite web versions:
// putting at the end ensures it overwrites in dist/cjs/index.js
export const View = WebView as any as RNTamaguiView
export const Stack = WebStack as any as RNTamaguiView
export const Text = WebText as any as RNTamaguiText

// easily test type declaration output and if it gets messy:

// export const X = styled(WebView, {
//   variants: {
//     abc: {
//       true: {},
//     },
//   } as const,
// })

// export const Y = styled(X, {
//   variants: {
//     zys: {
//       true: {},
//     },
//   } as const,
// })

// export const Z = styled(Y, {
//   variants: {
//     xxx: {
//       true: {},
//     },
//   } as const,
// })

// export const A = styled(Z, {
//   variants: {} as const,
// })

// const zz = <A />

// const variants = {
//   fullscreen: {
//     true: {},
//   },
//   elevation: {
//     '...size': () => ({}),
//     ':number': () => ({}),
//   },
// } as const

// export const YStack = styled(View, {
//   flexDirection: 'column',
//   variants,
// })

// import { TextInput } from 'react-native'
// export const InputFrame = styled(
//   TextInput,
//   {
//     name: 'Input',
//     backgroundColor: 'green',

//     variants: {
//       // unstyled: {
//       //   false: {},
//       // },

//       size: {
//         '...size': () => ({}),
//       },

//       // disabled: {
//       //   ':boolean': () => ({})
//       // },
//     } as const,

//     // defaultVariants: {
//     //   unstyled: process.env.TAMAGUI_HEADLESS === '1' ? true : false,
//     // },
//   },
//   {
//     isText: true,
//     accept: {
//       placeholderTextColor: 'color',
//     },
//   }
// )

// export const StyledInputFrame = styled(InputFrame, {
//   fontSize: 16,
//   fontFamily: '$silkscreen',
//   color: '$color5',
//   minWidth: 0,
//   borderWidth: 0,
//   borderColor: 'transparent',

//   variants: {
//     unset: {
//       false: {
//         borderWidth: 2,
//         py: 12,
//         px: 14,
//         borderRadius: 6,
//         bg: '$color3',
//         focusStyle: {
//           bg: '$color4',
//           margin: 0,
//         },
//       },
//     },
//   } as const,

//   defaultVariants: {
//     unset: false,
//   },
// })

// export const StyledStyledInputFrame = styled(
//   StyledInputFrame,
//   {
//     fontSize: 16,
//     fontFamily: '$silkscreen',
//     color: '$color5',
//     minWidth: 0,
//     borderWidth: 0,
//     borderColor: 'transparent',

//     variants: {
//       unset: {
//         false: {
//           borderWidth: 2,
//           py: 12,
//           px: 14,
//           borderRadius: 6,
//           bg: '$color3',
//           focusStyle: {
//             bg: '$color4',
//             margin: 0,
//           },
//         },
//       },
//     } as const,

//     defaultVariants: {
//       unset: false,
//     },
//   },
//   {
//     inlineProps: new Set(['id', 'testID']),
//   }
// )
// export const DepthTest = () => <StyledStyledInputFrame placeholder="" />
