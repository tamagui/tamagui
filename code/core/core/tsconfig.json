{
  "extends": "../../../tsconfig.json",
  "compilerOptions": {
    "composite": true,
    "strict": true
  },
  "exclude": [
    "./native.d.ts",
    "./native-test.d.ts",
    "./types",
    "types",
    "dist",
    "**/__tests__",
    "vite.bundle.config.ts"
  ],
  "references": [
    {
      "path": "../react-native-use-pressable"
    },
    {
      "path": "../react-native-use-responder-events"
    },
    {
      "path": "../use-event"
    },
    {
      "path": "../web"
    }
  ]
}
