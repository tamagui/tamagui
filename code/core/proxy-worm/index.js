module.exports = worm(true)

function worm(root = false) {
  const obj = root
    ? {
        get default() {
          return worm()
        },
      }
    : (() => {
        //  dev warning in debug mode
        if (
          process.env.NODE_ENV === 'development' &&
          process.env.DEBUG?.startsWith('tamagui')
        ) {
          console.warn(`

This has been excluded via Tamagui!
Check "excludeReactNativeWebExports" setting and include it to fix.

${new Error().stack}

`)
        }

        return worm()
      })

  obj.displayName = `ProxyWorm - Check excludeReactNativeWebExports`
  obj._isProxyWorm = true

  // reanimated tries to find component like things
  obj.prototype = obj.prototype || {}
  obj.prototype.isReactComponent = true

  return new Proxy(obj, {
    get(_, key) {
      if (Reflect.has(obj, key)) return Reflect.get(obj, key)
      return worm()
    },
    apply() {
      return worm()
    },
  })
}
