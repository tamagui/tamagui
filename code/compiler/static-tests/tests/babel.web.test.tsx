import * as React from 'react'
import { expect, test } from 'vitest'

import { extractForWeb } from './lib/extract'

window['React'] = React

test('theme props get extracted properly', async () => {
  const output = await extractForWeb(
    `
import { View } from '@tamagui/core'
    export function Test(props) {
      return (
        <View theme="green" width={10} bg={props.green ? 'red' : 'blue'} />
      )
    }
  `,
    {
      options: {
        platform: 'web',
        components: ['@tamagui/core'],
      },
    }
  )

  expect(output?.js).toContain(`<_TamaguiTheme name="green"><div className={`)
})

test('theme + media queries + conditionals extract', async () => {
  const output = await extractForWeb(
    `
    import { Stack } from '@tamagui/core'
    export function Test(props) {
      return (
        <Stack
          theme="surface1"
          $sm={{ flexDirection: 'column' }}
          {...(onlyDemo && {
            flexDirection: 'column',
          })}
        />
      )
    }
  `,
    {
      options: {
        platform: 'web',
        components: ['@tamagui/core'],
      },
    }
  )

  expect(output?.js).toMatchSnapshot()
  expect(output?.styles).toMatchSnapshot()
})

// floating && floating2 && p2 pb18 mr2 btrr10 br5 btlr7
// floating && !floating2 && p2 pb18 mr1 btrr10 br5
// !floating && floating2 && p2 pb15 mr2 btrr10 br2 btlr7
// !floating && !floating2 && p2 pb15 mr1 btrr10 br2

test('conditional specific after generic style overrides', async () => {
  const output = await extractForWeb(
    `
    import { Stack } from '@tamagui/core'
    export function Test(props) {
      return (
        <Stack
          p="$2"                              // base padding
          pb={floating ? 18 : 15}             // should override bottom
          mr={floating2 ? 2 : 1}              // unrelated ternary
          borderTopRightRadius={10}           // base tr radius
          borderRadius={floating ? 5 : 2}     // should override the tr radius always
          {...floating2 && {
            borderTopLeftRadius: 7
          }}
        />
      )
    }
  `,
    {
      options: {
        platform: 'web',
        components: ['@tamagui/core'],
      },
    }
  )

  expect(output?.js).toMatchSnapshot()
  expect(output?.styles).toMatchSnapshot()
})

test('conditional styles get full base styles merged onto + shorthand', async () => {
  const output = await extractForWeb(
    `
import { Stack } from '@tamagui/core'
    export function Test(props) {
      return (
        <Stack width={10} bg={props.green ? 'red' : 'blue'} />
      )
    }
  `,
    {
      options: {
        platform: 'web',
        components: ['@tamagui/core'],
      },
    }
  )

  expect(output?.js).toMatchSnapshot()
  expect(output?.styles).toMatchSnapshot()
})

test('className + conditional styles get full base styles merged onto + shorthand', async () => {
  const output = await extractForWeb(
    `
    import { Stack } from '@tamagui/core'
    export function Test(props) {
      return (
        <Stack width={10} bg={props.green ? 'red' : 'blue'} className={props.className} />
      )
    }
  `,
    {
      options: {
        platform: 'web',
        components: ['@tamagui/core'],
      },
    }
  )
  expect(output?.js).toMatchSnapshot()
  expect(output?.styles).toMatchSnapshot()
})

test('font classNames are extracted properly', async () => {
  // one sanity check debug output test
  const output = await extractForWeb(
    `
    import { Text } from '@tamagui/core'
    export function Test(props) {
      return (
        <Text fontFamily="$body" />
      )
    }
  `,
    {
      options: {
        platform: 'web',
        components: ['@tamagui/core'],
      },
    }
  )

  expect(
    output?.js.includes(
      `_cn = "font_body _dsp-inline _bxs-border-box _ww-break-word _whiteSpace-pre-wrap _mt-0px _mr-0px _mb-0px _ml-0px _ff-f-family"`
    )
  )
})

test('ternaries + font families works', async () => {
  // one sanity check debug output test
  const output = await extractForWeb(
    `
    import { Text } from '@tamagui/core'
    export function Test(props) {
      return (
        <Text fontFamily={window ? "$body" : "$heading"} />
      )
    }
  `,
    {
      options: {
        platform: 'web',
        components: ['@tamagui/core'],
      },
    }
  )

  expect(output?.js).toMatchSnapshot()
  expect(output?.styles).toMatchSnapshot()
})

test('bails from non-deterministic values', async () => {
  // one sanity check debug output test
  const output = await extractForWeb(
    `
    import { Text } from '@tamagui/core'
    export function Test(props) {
      return (
        <Text
          color={Math.random()}
        />
      )
    }
  `,
    {
      options: {
        platform: 'web',
        components: ['@tamagui/core'],
      },
    }
  )

  expect(output?.js).toMatchSnapshot()
})

test('non-flattened works', async () => {
  const output = await extractForWeb(
    `
    import { Text } from '@tamagui/core'
    export function Test(props) {
      return (
        <Text
          textAlign={inMenu ? props.hello : 'right'}
          width="100%"
          hoverStyle={{
            o: 0.85,
          }}
          {...(active && {
            fow: '700',
            opacity: 1,
          })}
        />
      )
    }
  `,
    {
      options: {
        platform: 'web',
        components: ['@tamagui/core'],
      },
    }
  )

  expect(output?.js).toMatchSnapshot()
})

test('fontFamily shorthand + styled + flatten works', async () => {
  const output = await extractForWeb(
    `
    import { MySizableText } from '@tamagui/test-design-system'
    export function Test(props) {
      return (
        <MySizableText
          ff="$mono"
        />
      )
    }
  `
  )

  expect(output?.js).toMatchSnapshot()
})

test('fontFamily shorthand + styled + flatten + ternaries', async () => {
  const output = await extractForWeb(
    `
    import { MySizableText } from '@tamagui/test-design-system'
    export function Test(props) {
      return (
        <MySizableText
          ff="$mono"
          opacity={active ? 1 : 0.65}
        />
      )
    }
  `
  )

  expect(output?.js).toMatchSnapshot()
})

test('specific className + ternary', async () => {
  const output = await extractForWeb(
    `
    import { MySizableText } from '@tamagui/test-design-system'
    export function Test(props) {
      return (
        <MySizableText
          className="test-class-name"
          ff="$mono"
          opacity={active ? 1 : 0.65}
        />
      )
    }
  `
  )

  expect(output?.js).toMatchSnapshot()
})

test('spread + className', async () => {
  const output = await extractForWeb(
    `
    import { View } from '@tamagui/core'

    export function Test(props) {
      return (
        <View
          className="ease-in-out-top"
          backgroundColor="#000"
          paddingVertical={2}
          top={0}
          {...(props.conditional && {
            top: -14,
            backgroundColor: '#fff',
          })}
        >
          {child}
        </View>
      )
    }
  `
  )

  expect(output?.js).toMatchSnapshot()
})

test('double ternary + spread', async () => {
  const output = await extractForWeb(
    `
    import { View } from '@tamagui/core'

    export function Test({ isSettings, isVertical, children }) {
      return (
        <View
          flex={isSettings || isVertical ? 'unset' : 5}
          alignItems="center"
          {...(isVertical && {
            flexDirection: 'column',
            flex: 'unset',
            alignItems: 'flex-start',
          })}
        >
          {children}
        </View>
      )
    }
  `
  )

  expect(output?.js).toMatchSnapshot()
})

test(`conditional classname keeps base and concats properly`, async () => {
  const output = await extractForWeb(
    `
    import { View } from '@tamagui/core'

    export function Test(props) {
      return (
        <View
          flex={1}
          className={isEnabled ? '' : 'disable-all-pointer-events'}
        >
          {props.child}
        </View>
      )
    }
  `
  )

  expect(output?.js).toMatchSnapshot()
})
