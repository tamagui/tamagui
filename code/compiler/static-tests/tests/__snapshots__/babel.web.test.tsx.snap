// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`className + conditional styles get full base styles merged onto + shorthand 1`] = `
"const _cn3 = " _bg-blue";
const _cn2 = " _bg-red";
const _cn = " _dsp-flex _ai-stretch _fd-column _fb-auto _bxs-border-box _pos-static _mih-0px _miw-0px _fs-0 _w-10px  ";
import { Stack } from '@tamagui/core';
export function Test(props) {
  return <div className={concatClassName("" + (_cn + (props.green ? _cn2 : _cn3) + (props.className ? " " + props.className : "")))} />;
}
import { concatClassName } from "@tamagui/helpers";"
`;

exports[`className + conditional styles get full base styles merged onto + shorthand 2`] = `
":root ._dsp-flex{display:flex;}
:root ._ai-stretch{align-items:stretch;}
:root ._fd-column{flex-direction:column;}
:root ._fb-auto{flex-basis:auto;}
:root ._bxs-border-box{box-sizing:border-box;}
:root ._pos-static{position:static;}
:root ._mih-0px{min-height:0px;}
:root ._miw-0px{min-width:0px;}
:root ._fs-0{flex-shrink:0;}
:root ._w-10px{width:10px;}
:root ._bg-red{background-color:red;}
:root ._bg-blue{background-color:blue;}"
`;

exports[`conditional styles get full base styles merged onto + shorthand 1`] = `
"const _cn3 = " _bg-blue";
const _cn2 = " _bg-red";
const _cn = " _dsp-flex _ai-stretch _fd-column _fb-auto _bxs-border-box _pos-static _mih-0px _miw-0px _fs-0 _w-10px  ";
import { Stack } from '@tamagui/core';
export function Test(props) {
  return <div className={concatClassName("" + (_cn + (props.green ? _cn2 : _cn3)))} />;
}
import { concatClassName } from "@tamagui/helpers";"
`;

exports[`conditional styles get full base styles merged onto + shorthand 2`] = `
":root ._dsp-flex{display:flex;}
:root ._ai-stretch{align-items:stretch;}
:root ._fd-column{flex-direction:column;}
:root ._fb-auto{flex-basis:auto;}
:root ._bxs-border-box{box-sizing:border-box;}
:root ._pos-static{position:static;}
:root ._mih-0px{min-height:0px;}
:root ._miw-0px{min-width:0px;}
:root ._fs-0{flex-shrink:0;}
:root ._w-10px{width:10px;}
:root ._bg-red{background-color:red;}
:root ._bg-blue{background-color:blue;}"
`;
